main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
+    case Some(path) => MultiLogger(logger.colored, log, new FileLogger(logger.colored, path, debugEnabled = true))
   }
+
+  type Terminal = Either[Task[_], Labelled[Any]]
+  type TerminalGroup = (Terminal, Agg[Task[_]])
+
+  def evaluateParallel(
+    goals: Agg[Task[_]],
+    threadCount: Int,
+    reporter: Int => Option[BuildProblemReporter] = (int: Int) => Option.empty[BuildProblemReporter],
+    testReporter: TestReporter = DummyTestReporter,
+    logger: Logger = log
+  ): Evaluator.Results = {
+    os.makeDir.all(outPath)
+
+    val startTime = System.currentTimeMillis()
+
+    // we need to collect all relevant logging info while developing
+    val evalLog = MultiLogger(true, this.log,
+      new FileLogger(false, outPath / "evaluator.log", true, append = true) {
+        override def debug(s: String) = super.debug(s"${System.currentTimeMillis() - startTime} [${Thread.currentThread().getName()}] ${s}")
+      })
+
+    evalLog.info(s"Using experimental parallel evaluator with ${threadCount} threads")
+    evalLog.debug(s"Start time: ${new java.util.Date()}")
+
+    val (sortedGroups, transitive) = Evaluator.plan(rootModule, goals)
+
+    // Mutable collector for all evaluated tasks
+    val evaluated = new Agg.Mutable[Task[_]]
+
+    // Mutable collector for all task results
+    @volatile var results = Map.empty[Task[_], Result[(Any, Int)]]
+
+    type Timing = (Either[Task[_], Labelled[_]], Int, Boolean)
+
+    case class FutureResult(task: TerminalGroup, time: Int, result: Evaluated)
+
+    // Mutable collector for timings
+    val timings = new mutable.ArrayBuffer[Timing](sortedGroups.keyCount)
+
+    // Increment the counter message by 1 to go from 1/10 to 10/10
+    val nextCounterMsg = new Evaluator.NextCounterMsg(sortedGroups.keyCount)
+
+    // TODO: check for interactivity
+    // TODO: make sure, multiple goals run in order, e.g. clean compile
+
+    val executorService = Executors.newFixedThreadPool(threadCount)
+    evalLog.debug(s"Created executor: ${executorService}")
+    val completionService = new ExecutorCompletionService[FutureResult](executorService)
+
+    // The scheduled and not yet finished futures (Java!)
+    var futures = List[(java.util.concurrent.Future[FutureResult], TerminalGroup)]()
+
+    try {
+
+      val interGroupDeps: Map[TerminalGroup, Seq[TerminalGroup]] = findInterGroupDeps(sortedGroups)
+      evalLog.debug(s"${interGroupDeps} (took ${System.currentTimeMillis() - startTime} msec)")
+
+      // State holders, only written to from same thread
+      // The unprocessed terminal groups
+      var work = sortedGroups.items().toList
+      // The currently scheduled (maybe not started yet) terminal groups
+      var inProgress = List[TerminalGroup]()
+      // The finished terminal groups
+      var doneMap = Map[TerminalGroup, Boolean]().withDefaultValue(false)
+      // The fact that at least one task failed
+      @volatile var someTaskFailed: Boolean = false
```

Comment:
This only seems to be read/written from the main thread; does it really need to be `@volatile`?

Reply:
It's also read from inside a future when failFast is enabled.

Reply:
Got it. If we move the body of the future to a different method, we can make this an `AtomicBoolean` to serve the same purpose

Reply:
Done